extensions/amp-form/0.1/amp-form.js

```diff
@@ -258,6 +261,20 @@ export class AmpForm {
     return url;
   }
 
+  /**
+   * Gets and validates an attribute for form request encoding type.
+   * @param {string} attribute
+   * @return {?string}
+   * @private
+   */
+  getEncType_(attribute) {
+    const encType = this.form_.getAttribute(attribute);
+    if (encType === 'application/x-www-form-urlencoded') {
+      return encType;
+    }
```

Comment:
Should we emit a `user().warn` warning in case they've attempted to use an unsupported encType.

Reply:
Thanks for the suggestion Jake.  I have included a user().warn() statement when the user specifies an unsupported enctype.  

I saw the the warning shows up in the console, but is it logged anywhere else?  Also is there somewhere I can get a better understanding of the overall logging process.  I see that there are a few other methods besides warn() in the log.js file and I'd like to get a better understanding of how / when to use each.  Thanks!!

fyi: @caroqliu

Reply:
I don't personally have a great handle on all of the logging (and error) flows. I'll just word-vomit some things here that I think I do know (but please correct me @caroqliu if I get anything wrong).

There are two categories of logs: `user` and `dev`. For the most part, they behave similarly. `dev` warnings are meant to be consumed by developers of the amp framework (us), whereas the `user` warnings are meant to be consumed by developers using AMP (publishers).

Within logging (and in many logging frameworks in other languages/libraries as well), there are many logging levels. I think we have between 1 and 4.  When loading an amp page, you can enter debug mode by adding `#developement=1` to the url and you can also specify the log level with `#log=N` (where N is either 0,1,2,3 or 4). You could also type something like `AMP.setLogLevel` in the console.

Logs are only displayed if they are a lower level than the log level setting (aka a higher priority log).

Now in terms of special gotchyas, you can check `src/build-system/babel-plugins/babel-plugin-transform-amp-asserts`. There are certain things that only exist at development time but are deleted for production builds. Here is the list:
```js
const REMOVABLE = {
  dev: [
    'assert',
    'fine',
    'assertElement',
    'assertString',
    'assertNumber',
    'assertBoolean',
    'assertArray',
  ],
  user: ['fine'],
};
```

This means that `fine()` level logs are great for in-depth debugging info since they will never get into the production build and increase our bundle sizes.

One more gotchya is that user errors and dev errors are all sent to an external service. We randomly decide to skip sending certain errors based on a probability, and I believe that we cut out 100x more user() errors.

Reply:
Thanks for that thorough context Jake, extremely helpful!

One quick follow up based on the list you provided.  Does this mean that

devAssert
devFine
devAssertElement, etc..

Are removed from production builds?  Or are those a log level?
Thanks!

Reply:
I'm pretty sure those are removed as well, but I haven't checked to see if that happens within this babel transform or a different one. I just figure it would be silly to remove `dev().assert()` and not remove its equivalent and preferred `devAssert`